Lib/datetime.py

```diff
@@ -692,9 +692,12 @@ def __new__(cls, year, month=None, day=None):
 
         year, month, day (required, base 1)
         """
-        if month is None and isinstance(year, bytes) and len(year) == 4 and \
-                1 <= year[2] <= 12:
+        if (month is None and
+            isinstance(year, (bytes, str)) and len(year) == 4 and
+            1 <= ord(year[2:3])&0xFF <= 12):
```

Comment:
If the third byte is in the range 0x00-0x7f, it is decoded to the character in the range U+0000-U+007f. If it is in the range 0x80-0xff, it is decoded to the character in the range U+dc80-U+dcff.

Hmm, but in the latter case `ord(year[2:3])&0xFF >= 0x80`, and that condition is false. `&0xFF` is not needed in this case. But '&0x7F' in other two cases are needed.